07_VirtualFileSystem/02_VirtualFileSystem.md

```diff
@@ -0,0 +1,433 @@
+
+```c
+struct fs_operations_t {
+	int (*open)(const char *path, int flags, ... );
+	int (*close)(int file_descriptor);
+	ssize_t (*read)(int file_descriptor, char* read_buffer, size_t nbyte);
+	ssize_t (*write)(int file_descriptor, const void* write_buffer, size_t nbyte);
+};
+
+typedef struct fs_operations_t fs_operations_t;
+```
+
+The basic idea is that once the mountpoint_id has been found (we declared `get_mountpoint_id` in the previous paragraph), the vfs will use the mountpoint item to call the fs driver implementation of the open function, but remember that when calling the driver function, it doesn't care about the full path, it only needs the relative path with mountpoint folder stripped, and if the whole path is passed, we will most likely get an error. Since the fs root will start from within the mountpoint folder we need to get the relative path, we will use the `get_rel_path` function defined earlier in this chapter, and the pseudocode for the open function should look similar to the following:
+
+
+```c
+int open(const char *path, int flags){
+    mountpoint_id = get_mountpoint_id(pathname);
+    
+    if (mountpoint_id > -1) {
+        char *rel_path = get_rel_path(mountpoints[mountpoint_id], path);
+        int fs_specific_id = mountpoints[mountpoint_id].operations.open(rel_path, flags);
+        if (fs_specific_id != ERROR) {
+            /* IMPLEMENTATION LEFT AS EXERCISE */
+            // Get a new vfs descriptor id vfs_id
+            vfs_opened_files[vfs_id] = //fill the file descriptor entry at position 
+        }
+    }
+    return vfs_id
+}
+```
+
+The pseudo code above should give us an idea of what is the workflow of opening a file from a VFS point of view, as you can see the process is pretty simple in principle: getting the mountpoint_id from the vfs, if one has been found get strip out the mountpoint path from the path name, and call the fs driver open function, if this function call is succesfull is time to initialize a new vfs file descriptor item. 
+
+The `close` with signature: 
+
+```c 
+int close(int fildes);
+```
+
+will take a vfs file descriptor as parameter, and will search for it in the opened files list (using an array it will be found at `vfs_opened_files[fildes]`) and if found it should first call the fs driver function to close a file (if present), emptying all data structures associated to that file descriptor (i.e. if there are data on pipes or FIFO they should be discarded) and finally it can mark this position as available again. We have only one problem how to mark a file descriptor available using an array? One idea can be to use -1 as `fs_file_id` to identify a position that is marked as available (so we will need to set them to -1 when the vfs is initialized).
```

Comment:
This feels a bit janky starting a sentence like this. Maybe move the code above the paragraph, and then reference it by starting with something like "The example close() function shown above will take a vfs file ..."